src/main/java/com/google/sps/DataServlet.java

```diff
@@ -169,16 +161,38 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     }
     truncatedGraph = currDataGraph.getReachableNodes(nodeNameParam, depthNumber);
 
-    // We set the headers in the following 3 scenarios:
+    if (nodeNameParam.equals("") && truncatedGraph.equals(currDataGraph.graph())) {
+      // If we are not filtering the graph or limiting its depth, show all mutations of all nodes
+      filteredMutationIndices = defaultIndices;
+    } else {
+      // Get the names of all the displayed nodes and find all indices of mutations
+      // that mutate any of them
+      Set<String> truncatedGraphNodeNames = Utility.getNodeNamesInGraph(truncatedGraph);
+      // Also get mutations relevant to the searched node if it is not an empty string
+      if (nodeNameParam.length() != 0) {
+        truncatedGraphNodeNames.add(nodeNameParam);
+      }
+      filteredMutationIndices =
+          Utility.findRelevantMutations(truncatedGraphNodeNames, mutationIndicesMap, mutList);
+
+      // Filter the diff to only show mutations relevant to the above nodes
+      diff = Utility.filterMultiMutationByNodes(diff, truncatedGraphNodeNames);
+    }
+
+    // We set the headers in the following 4 scenarios:
     // The searched node is not in the graph and is never mutated
     if (truncatedGraph.nodes().size() == 0 && filteredMutationIndices.size() == 0) {
       response.setHeader(
           "serverError", "The searched node does not exist anywhere in this graph or in mutations");
       return;
     }
     // The searched node is not in the graph but is mutated at some past/future
-    // point
-    if (truncatedGraph.nodes().size() == 0 && filteredMutationIndices.size() != 0) {
+    // point. The diff conditions are included to prevent entry into this case
+    // when the searched node is deleted for example. The diff being non-empty
+    // means that there is some mutation pertaining to the searched node to show
+    if (truncatedGraph.nodes().size() == 0
```

Comment:
Is there any way to make this expression account for deleted nodes?

Reply:
Oops I think I removed the extra diff check mentioned in the comment thinking that it was redundant.